File: backend/controllers/patientController.js
const Patient = require('../models/Patient');
const User = require('../models/User');

const updatePatientProfile = async (req, res) => {
  const { name, email, age, contact } = req.body;

  try {
    let patient = await Patient.findOne({ userId: req.user.id });

    if (!patient) {
      patient = await Patient.create({
        userId: req.user.id,
        name,
        email,
        age,
        contact,
      });
    } else {
      patient.name = name || patient.name;
      patient.email = email || patient.email;
      patient.age = age || patient.age;
      patient.contact = contact || patient.contact;
      await patient.save();
    }

    res.status(200).json(patient);
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};

const getPatientProfile = async (req, res) => {
  try {
    const patient = await Patient.findOne({ userId: req.user.id });

    if (!patient) {
      return res.status(404).json({ message: 'Patient profile not found' });
    }

    res.status(200).json(patient);
  } catch (error) {
    res.status(500).json({ message: 'Server error', error: error.message });
  }
};


const getAllPatients = async (req, res) => {
  try {
    const patients = await User.find({ role: 'patient' }).select('_id name email');
    res.json(patients);
  } catch (err) {
    res.status(500).json({ message: 'Failed to fetch patients', error: err.message });
  }
};

module.exports = { updatePatientProfile, getPatientProfile,getAllPatients };

File: backend/controllers/patient_controller.py
from flask import request, jsonify
from models.patient import Patient
from models.user import User


# ✅ Create or update patient profile
def update_patient_profile():
    data = request.get_json()
    user_id = getattr(request, "user", {}).get("id")

    try:
        patient = Patient.objects(userId=user_id).first()

        if not patient:
            patient = Patient(
                userId=user_id,
                name=data.get("name"),
                email=data.get("email"),
                age=data.get("age"),
                contact=data.get("contact")
            )
        else:
            patient.name = data.get("name", patient.name)
            patient.email = data.get("email", patient.email)
            patient.age = data.get("age", patient.age)
            patient.contact = data.get("contact", patient.contact)

        patient.save()
        return jsonify(patient.to_mongo().to_dict()), 200

    except Exception as error:
        return jsonify({"message": "Server error", "error": str(error)}), 500


# ✅ Get the currently logged-in patient's profile
def get_patient_profile():
    user_id = getattr(request, "user", {}).get("id")

    try:
        patient = Patient.objects(userId=user_id).first()

        if not patient:
            return jsonify({"message": "Patient profile not found"}), 404

        return jsonify(patient.to_mongo().to_dict()), 200

    except Exception as error:
        return jsonify({"message": "Server error", "error": str(error)}), 500


# ✅ Get all patients from User collection (role = 'patient')
def get_all_patients():
    try:
        patients = User.objects(role='patient').only('id', 'name', 'email')
        return jsonify([
            {
                "id": str(p.id),
                "name": p.name,
                "email": p.email
            } for p in patients
        ])

    except Exception as err:
        return jsonify({"message": "Failed to fetch patients", "error": str(err)}), 500
